File: apps/java/design-pattern/src/test/java/pattern/adapter/AdapterTest.java
package pattern.adapter;

import org.junit.jupiter.api.Test;

import static org.junit.jupiter.api.Assertions.*;

class AdapterTest {

    @Test
    void textObjectHoldsTextSizeAndColor() {
        TextObject obj = new TextObject("Hello", 2.0, "red");

        assertEquals("Hello", obj.getText());
        assertEquals(2.0, obj.getSizeInches());
        assertEquals("red", obj.getColor());
    }

    @Test
    void adapterConvertsBritishTextObjectToTextObject() {
        BritishTextObject bto = new BritishTextObject("Cheerio", 25.4, "grey");
        TextObject adapted = new BritishTextObjectAdapter(bto);

        assertEquals("Cheerio", adapted.getText());
        assertEquals(1.0, adapted.getSizeInches(), 0.001);
        assertEquals("grey", adapted.getColor());
    }

    @Test
    void rendererWorksWithAdaptedObject() {
        BritishTextObject bto = new BritishTextObject("Hello", 50.8, "blue");
        TextObject adapted = new BritishTextObjectAdapter(bto);
        Renderer renderer = new Renderer();

        String output = renderer.render(adapted);

        assertEquals("text:Hello size:2.0 color:blue", output);
    }

    @Test
    void rendererWorksWithNativeTextObject() {
        TextObject obj = new TextObject("Native", 3.5, "green");
        Renderer renderer = new Renderer();

        String output = renderer.render(obj);

        assertEquals("text:Native size:3.5 color:green", output);
    }
}

File: apps/scala/design-pattern/test/AdapterTest.scala
package designpattern.adapter

class AdapterSuite extends munit.FunSuite:
  test("BritishTextObjectAdapter が TextObject インターフェースを提供する") {
    val british = BritishTextObject()
    val adapter = BritishTextObjectAdapter(british)

    assertEquals(adapter.text, "default british text")
    assertEquals(adapter.sizeInBytes, 20) // lengthInChars
  }

  test("アダプタ経由でテキストを変更する") {
    val british = BritishTextObject()
    val adapter = BritishTextObjectAdapter(british)
    adapter.text = "hello"

    assertEquals(adapter.text, "hello")
    assertEquals(british.string, "hello")
  }

  test("UTF-8 アダプタが日本語のバイト数を正しく返す") {
    val british = BritishTextObject()
    british.string = "こんにちは"
    val adapter = Utf8BritishTextObjectAdapter(british)

    assertEquals(adapter.text, "こんにちは")
    assertEquals(adapter.sizeInBytes, 15) // UTF-8 で日本語は 3 バイト/文字
  }

  test("TextObject trait を満たすことをコンパイル時に保証する") {
    val british = BritishTextObject()
    val obj: TextObject = BritishTextObjectAdapter(british)

    assert(obj.isInstanceOf[TextObject])
  }
